File: sonia_common_py/sonia_common_py/serial_conn.py
"""
This module allows us to connect to Serial Devices in python.
"""
__all__ = ['SerialConn']

import fcntl
import os
import termios
from enum import Enum
from fcntl import fcntl
from termios import tcgetattr, tcsetattr
from typing import Union, List
from .execptions.exception_codes import SoniaCommonExceptionCodes
from .execptions.sonia_common_exception import SoniaCommonException

class SerialConn:
    """
    Serial port connection Manager.
    """

    BUFFER_SIZE: int = 1024
    """Size of a packet."""

    class __TermiosFlags(Enum):
        I_FLAG = 0
        O_FLAG = 1
        C_FLAG = 2
        L_FLAG = 3
        I_SPEED = 4
        O_SPEED = 5
        CC = 6

    def __init__(self, port: str, baud: int, is_blocking: bool = True) -> None:
        """
        Constructor.

        Args:
            port (str): COMX or tty/ depending on platform
            baud (int): baud rate. Use constants from termios module.
            is_blocking (bool, optional): Is the port connection blocking. Defaults to True.

        Raises:
            SoniaCommonException: Bad Baud rate.
        """
        self.__port = port
        if self.__check_baud(self.__check_baud(baud)):
            self.__baud = baud
        else:
            raise SoniaCommonException(SoniaCommonExceptionCodes.BAD_BAUD_RATE)
        self.__is_blocking = is_blocking
        self.__fd = None
        self.__options: List[int | list[bytes | int]] = None

    def __del__(self) -> None:
        """Destructor."""
        os.close(self.__fd)

    def read_packets(self, count: int, data: bytearray) -> int:
        """
        Read packets from the serial port.

        Args:
            count (int): number of packets of size BUFFER_SIZE.
            data (bytearray): Empty array that will be filled with the data from the port.

        Returns:
            int: 0 if successful else error code.
        """
        data = os.read(self.__fd, count * self.BUFFER_SIZE)
        return self.__validate_read(data, count)

    def read_once(self, data: bytearray, append: bool = False) -> int:
        """
        Read one packet of BUFFER_SIZE from the port.

        Args:
            data (bytearray): empty byte array to be filled.
            append (bool, optional): Append to existing data or override. Defaults to False.

        Returns:
            int: 0 for success else error code.
        """
        old_len = len(data) / self.BUFFER_SIZE
        if append:
            data.append(os.read(self.__fd, self.BUFFER_SIZE))
            return self.__validate_read(data, old_len + 1)

        data = os.read(self.__fd, self.BUFFER_SIZE)
        return self.__validate_read(data, old_len + 1)

    def flush(self) -> None:
        """
        Flush the port's IO stream.
        """
        termios.tcflush(self.__fd, termios.TCIOFLUSH)

    def transmit(self, data: Union[str, bytearray]) -> int:
        """
        Write to the port.

        Args:
            data (Union[str, bytearray]): Message to send.

        Returns:
            int: 0 for success else error code.
        """
        return os.write(self.__fd, data)

    def open_port(self) -> bool:
        """
        Open the port.

        Returns:
            bool: True for success else False.
        """
        if self.__is_blocking:
            self.__fd = os.open(self.__port, os.O_RDWR | os.O_NOCTTY)
        else:
            self.__fd = os.open(self.__port, os.O_RDWR |
                                os.O_NOCTTY | os.O_NDELAY)

        if self.__fd == -1:
            return False

        if not self.__is_blocking:
            fcntl(self.__fd, fcntl.F_SETFL, os.O_NDELAY)

        self.__options = tcgetattr(self.__fd)
        self.__options[self.__TermiosFlags.I_SPEED] = self.__baud
        self.__options[self.__TermiosFlags.O_SPEED] = self.__baud

        self.__options[self.__TermiosFlags.C_FLAG] |= (
            termios.CLOCAL | termios.CREAD)
        self.__options[self.__TermiosFlags.C_FLAG] &= ~termios.CSIZE
        self.__options[self.__TermiosFlags.C_FLAG] |= termios.CS8

        self.__options[self.__TermiosFlags.C_FLAG] &= ~(
            termios.PARENB | termios.PARODD)
        self.__options[self.__TermiosFlags.C_FLAG] &= ~termios.CSTOP
        self.__options[self.__TermiosFlags.C_FLAG] &= ~termios.CRTSCTS

        self.__options[self.__TermiosFlags.I_FLAG] &= ~termios.IGNBRK
        self.__options[self.__TermiosFlags.I_FLAG] &= ~(
            termios.IXON | termios.IXOFF | termios.IXANY)

        self.__options[self.__TermiosFlags.L_FLAG] = 0

        self.__options[self.__TermiosFlags.O_FLAG] = 0

        tcsetattr(self.__fd, termios.TCSANOW, self.__options)

        return True

    def __validate_read(self, data, count) -> int:
        if data is None:
            return -1
        if len(data < count * self.BUFFER_SIZE):
            return 1
        return 0

    @staticmethod
    def __check_baud(baud: int) -> bool:
        return baud in [
            eval(f"termios.{x}") for x in dir(termios) if x[0] == 'B' and x[1].isnumeric() #pylint: disable=eval-used
            ]

File: sonia_common_cpp/src/SerialConn.cpp
#include "SerialConn.hpp"

#include <fcntl.h>
#include <unistd.h>

namespace sonia_common_cpp
{
    SerialConn::SerialConn(std::string port, speed_t baud, uint8_t timeout = 5) : _port(port), _baud(baud), _isBlocking(true), _timeout(timeout) {}

    SerialConn::SerialConn(std::string port, speed_t baud, bool isBlocking, uint8_t timeout = 5)
        : _port(port), _baud(baud), _isBlocking(isBlocking), _timeout(timeout)
    {}

    SerialConn::~SerialConn()
    {
        // Close port on destruction.
        close(_fd);
    }

    ssize_t SerialConn::ReadPackets(size_t count, uint8_t *pData)
    {
        pData[0] = 0;
        fd_set set;
        struct timeval timeout;

        FD_ZERO(&set);
        FD_SET(_fd, &set);

        timeout.tv_sec = _timeout;
        timeout.tv_usec = 0;
        
        int rv;
        _lock.lock();
        rv = select(_fd + 1, &set, NULL, NULL, &timeout);
        if (rv == -1) {
            _lock.unlock();
            return -1; // Error
        } else if (rv == 0) {
            _lock.unlock();
            return 0; // Timeout, no data
        }
        ssize_t ret = read(_fd, pData, count);
        _lock.unlock();
        return ret;
    }

    ssize_t SerialConn::ReadOnce(uint8_t *pData, int offset)
    {
        _lock.lock();
        ssize_t ret = read(_fd, (pData + offset), 1);
        _lock.unlock();
        return ret;
    }

    void SerialConn::Flush() const { tcflush(_fd, TCIOFLUSH); }

    ssize_t SerialConn::Transmit(const std::string data)
    {
        _lock.lock();
        ssize_t ret = write(_fd, data.c_str(), data.size());
        _lock.unlock();
        return ret;
    }

    ssize_t SerialConn::Transmit(const uint8_t *pData, size_t length)
    {
        _lock.lock();
        ssize_t ret = write(_fd, pData, length);
        _lock.unlock();
        return ret;
    }

    bool SerialConn::OpenPort()
    {
        /*
        O_RDWR: Read and write
        O_NOCTTY: No control TTY
        O_NDELAY: Non-blocking mode.
        */
        if (_isBlocking)
        {
            _fd = open(_port.c_str(), O_RDWR | O_NOCTTY);
        }
        else
        {
            _fd = open(_port.c_str(), O_RDWR | O_NOCTTY | O_NDELAY);
        }

        if (_fd == -1)
        {
            return false;
        }

        if (!_isBlocking)
        {
            fcntl(_fd, F_SETFL, O_NDELAY);
        }

        // Get the options object from the port.
        tcgetattr(_fd, &_options);

        // Set the baud rate
        cfsetispeed(&_options, _baud);
        cfsetospeed(&_options, _baud);

        // Hardware control of the terminal
        // flags defined here: https://www.ibm.com/docs/en/aix/7.1?topic=files-termiosh-file
        _options.c_cflag |= (CLOCAL | CREAD);  // Specify local ligne (local direct connection), Enables receiver
        _options.c_cflag &= ~CSIZE;            // Remove set character size
        _options.c_cflag |= CS8;               // Set character size to 8 bits

        _options.c_cflag &= ~(PARENB | PARODD);  // Disable parity, specify even paraity
        _options.c_cflag &= ~CSTOPB;   // remove specified number stop bits. This sets the default of 1 stop bit.
        _options.c_cflag &= ~CRTSCTS;  // Disable flow control

        // Input Flags
        _options.c_iflag &= ~IGNBRK;                  // Do not ignore break conditions
        _options.c_iflag &= ~(IXON | IXOFF | IXANY);  // disable start and stop output control, Disable start-and-stop
                                                      // input control, disable any char to restart output.

        // Local Flags
        _options.c_lflag = 0;

        // Output Flags
        _options.c_oflag = 0;

        // Set the new options to the port.
        tcsetattr(_fd, TCSANOW, &_options);
        return true;
    }
}  // namespace sonia_common_cpp
